src/ws/cert-session.c

```diff
@@ -0,0 +1,307 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This program opens a session for a given user and runs the bridge in
+ * it, using TLS client certificate authentication and mapping it to a user
+ * through sssd (https://www.freeipa.org/page/V4/User_Certificates).
+ */
+
+#include "config.h"
+#include <stdlib.h>
+
+#include <gio/gio.h>
+
+#include "session-utils.h"
+
+static char *last_txt_msg = NULL;
+static char *conversation = NULL;
+
+
+static int
+pam_conv_func (int num_msg,
+               const struct pam_message **msg,
+               struct pam_response **ret_resp,
+               void *appdata_ptr)
+{
+  /* we don't expect (nor can handle) any actual auth conversation here, but
+   * PAM sometimes sends messages like "Creating home directory for USER" */
+  for (int i = 0; i < num_msg; ++i)
+      debug("got PAM conversation message, ignoring: %s", msg[i]->msg);
+  return PAM_CONV_ERR;
+}
+
+static char* sssd_map_certificate (const char *certificate)
+{
+  g_autoptr(GDBusConnection) bus = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) user_obj_v = NULL;
+  const gchar *user_obj = NULL;
+  g_autoptr(GVariant) username_v = NULL;
+  g_autoptr(GVariant) username_vv = NULL;
```

Comment:
This is the first use of g_autoptr in the project. Are we okay with that?

Reply:
We recently bumped the glib dependency to a new enough version, so technically it works. The gcc cleanup stuff is being used in systemd and other things all over the place, so from my POV I don't have doubts about that.